compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingForall.cpp

```diff
@@ -0,0 +1,233 @@
+    llvm::SmallDenseSet<unsigned> partitionableLoopsSet(
+        partitionableLoops.begin(), partitionableLoops.end());
+    OpFoldResult zero = rewriter.getIndexAttr(0);
+    for (auto loopId : llvm::seq<unsigned>(0, tileSizes.size())) {
+      if (partitionableLoopsSet.count(loopId)) {
+        continue;
+      }
+      tileSizes[loopId] = zero;
+    }
+  }
+
+  return TilingInfo{tilableOp, tileSizes, interchange};
+}
+
+/// Helper function to return the mapping attribute to use given the tile sizes
+static SmallVector<Attribute> getMapping(MLIRContext *context,
+                                         ArrayRef<OpFoldResult> tileSizes) {
+  SmallVector<Attribute> mapping;
+  mapping.reserve(tileSizes.size());
+  for (auto tileSize : llvm::reverse(tileSizes)) {
+    if (isConstantIntValue(tileSize, 0)) {
+      continue;
+    }
+    uint64_t currSize = mapping.size();
+    switch (currSize) {
+    case 0:
+    case 1:
+    case 2:
+      mapping.push_back(IREE::Codegen::WorkgroupMappingAttr::get(
+          context, IREE::Codegen::symbolizeWorkgroupId(currSize).value()));
+      break;
+    default:
+      mapping.push_back(IREE::Codegen::WorkgroupMappingAttr::get(
+          context, IREE::Codegen::WorkgroupId::IdZ, currSize - 2));
+    }
+  }
+  return llvm::to_vector(llvm::reverse(mapping));
+}
+
+static SmallVector<int64_t> getStaticNumWorkgroups(RewriterBase &rewriter,
+                                                   scf::ForallOp loop) {
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(loop);
+
+  SmallVector<OpFoldResult> lowerBound = loop.getMixedLowerBound();
+  SmallVector<OpFoldResult> upperBound = loop.getMixedUpperBound();
+  SmallVector<OpFoldResult> step = loop.getMixedStep();
+
+  AffineExpr s0, s1, s2;
+  bindSymbols(rewriter.getContext(), s0, s1, s2);
+  AffineExpr numItersExpr = (s1 - s0).ceilDiv(s2);
+  SmallVector<OpFoldResult> numIters = llvm::map_to_vector(
+      llvm::zip_equal(lowerBound, upperBound, step), [&](auto it) {
+        return affine::makeComposedFoldedAffineApply(
+            rewriter, loop.getLoc(), numItersExpr,
+            {std::get<0>(it), std::get<1>(it), std::get<2>(it)});
+      });
+  SmallVector<Value> dynamicNumIters;
+  SmallVector<int64_t> staticNumIters;
+  dispatchIndexOpFoldResults(numIters, dynamicNumIters, staticNumIters);
+  return staticNumIters;
+}
+
+void TileAndDistributeToWorkgroupsUsingForallOpPass::runOnOperation() {
+  auto funcOp = getOperation();
+  auto *context = &getContext();
+  SmallVector<Operation *> computeOps = getComputeOps(funcOp);
+
+  IRRewriter rewriter(context);
+  FailureOr<TilingInfo> tilingInfo =
+      getTiledAndDistributionInfo(rewriter, computeOps);
+  if (failed(tilingInfo)) {
+    return signalPassFailure();
+  }
+  auto tilableOp = dyn_cast_or_null<TilingInterface>(tilingInfo->tilableOp);
+  if (!tilableOp) {
+    // Did not find a tileable op. So do nothing.
```

Comment:
What is expected to happen to the workgroup count region if we do nothing? Do we just populate it with `[1, 1, 1]` if no `scf.forall` with workgroup mapping is found?

Reply:
Yeah, that will be handled later on when we resolve `scf.forall` operations.